src/ripple/app/ledger/impl/DisputedTx.h

```diff
@@ -73,26 +72,161 @@ class DisputedTx
 };
 
+// Track a peer's yes/no vote on a particular disputed transaction
+template <class Tx_t, class TxID_t, class NodeID_t>
+void DisputedTx<Tx_t,TxID_t,NodeID_t>::setVote (NodeID_t const& peer, bool votesYes)
+{
+    auto res = mVotes.insert (std::make_pair (peer, votesYes));
+
+    // new vote
+    if (res.second)
+    {
+        if (votesYes)
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes YES on " << mTransactionID;
+            ++mYays;
+        }
+        else
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes NO on " << mTransactionID;
+            ++mNays;
+        }
+    }
+    // changes vote to yes
+    else if (votesYes && !res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes YES on " << mTransactionID;
+        --mNays;
+        ++mYays;
+        res.first->second = true;
+    }
+    // changes vote to no
+    else if (!votesYes && res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes NO on " << mTransactionID;
+        ++mNays;
+        --mYays;
+        res.first->second = false;
+    }
+}
+
+// Remove a peer's vote on this disputed transasction
+template <class Tx_t, class TxID_t, class NodeID_t>
+void DisputedTx<Tx_t,TxID_t,NodeID_t>::unVote (NodeID_t const& peer)
+{
+    auto it = mVotes.find (peer);
+
+    if (it != mVotes.end ())
+    {
+        if (it->second)
+            --mYays;
+        else
+            --mNays;
+
+        mVotes.erase (it);
+    }
+}
+
+template <class Tx_t, class TxID_t, class NodeID_t>
+bool DisputedTx<Tx_t,TxID_t,NodeID_t>::updateVote (int percentTime, bool proposing)
+{
+    // VFALCO TODO Give the return value a descriptive local variable name
+    //             and don't return from the middle.
+
+    if (mOurVote && (mNays == 0))
+        return false;
+
+    if (!mOurVote && (mYays == 0))
+        return false;
+
+    bool newPosition;
+    int weight;
+
+    if (proposing) // give ourselves full weight
+    {
+        // This is basically the percentage of nodes voting 'yes' (including us)
+        weight = (mYays * 100 + (mOurVote ? 100 : 0)) / (mNays + mYays + 1);
+
+        // VFALCO TODO Rename these macros and turn them into language
+        //             constructs.  consolidate them into a class that collects
+        //             all these related values.
+        //
+        // To prevent avalanche stalls, we increase the needed weight slightly
+        // over time.
+        if (percentTime < AV_MID_CONSENSUS_TIME)
+            newPosition = weight >  AV_INIT_CONSENSUS_PCT;
+        else if (percentTime < AV_LATE_CONSENSUS_TIME)
+            newPosition = weight > AV_MID_CONSENSUS_PCT;
+        else if (percentTime < AV_STUCK_CONSENSUS_TIME)
+            newPosition = weight > AV_LATE_CONSENSUS_PCT;
+        else
+            newPosition = weight > AV_STUCK_CONSENSUS_PCT;
```

Comment:
Good suggestion. I'll hold that over for the next set of changes. Otherwise, we'll wind up having the consensus tuning parameters that DisputedTx uses in one place and the ones it doesn't in another, which would make a bit of a mess.